File: apps/cargo-tracker/src/Cargotracker/Booking/Domain/Model/Value/RouteSpecification.hs
{- | 経路仕様 (IT1 US04)

出発港 / 到着港 / 到着期限の 3 つを束ねる値オブジェクト。
IT1 では構造のみ。同港間予約 (origin == destination) は許容し、
IT2 で経路設計時の `RouteSatisfied` 検証で弾く。
-}
module Cargotracker.Booking.Domain.Model.Value.RouteSpecification
  ( RouteSpecification (..),
  ) where

import Data.Time (UTCTime)

import Cargotracker.Shared.Domain.Common.UnLocode (UnLocode)

data RouteSpecification = RouteSpecification
  { origin :: !UnLocode
  , destination :: !UnLocode
  , arrivalDeadline :: !UTCTime
  }
  deriving stock (Eq, Show)

File: apps/cargo-tracker/packs/booking/app/domain/booking/domain/route_specification.rb
# frozen_string_literal: true

module Booking
  module Domain
    # ルート仕様。出発地・目的地（UN/LOCODE）・到着期限の要件。
    # 注: IT2 は UN/LOCODE を String で保持。Location 共有カーネル化は IT3。
    RouteSpecification = Data.define(:origin, :destination, :arrival_deadline) do
      def initialize(origin:, destination:, arrival_deadline:)
        raise ArgumentError, "出発地は必須です" if origin.nil? || origin.to_s.strip.empty?
        raise ArgumentError, "目的地は必須です" if destination.nil? || destination.to_s.strip.empty?
        raise ArgumentError, "出発地と目的地は異なる必要があります" if origin == destination
        raise ArgumentError, "到着期限は必須です" if arrival_deadline.nil?

        super
      end
    end
  end
end
